packages/browser/frontend/sirius-components-browser/src/modelbrowser/ModelBrowserTreeView.tsx

```diff
@@ -53,17 +62,106 @@ export const ModelBrowserTreeView = ({
     maxDepth: 1,
   });
 
+  const { getTreePath, loading: treePathLoading, data: treePathData } = useTreePath();
+  const {
+    getExpandAllTreePath,
+    loading: expandAllTreePathLoading,
+    data: expandAllTreePathData,
+  } = useExpandAllTreePath();
+
+  useEffect(() => {
+    if (!expandAllTreePathLoading) {
+      if (expandAllTreePathData) {
+        const { expanded, maxDepth } = state;
+        if (expandAllTreePathData.viewer?.editingContext?.expandAllTreePath) {
+          const { treeItemIdsToExpand, maxDepth: expandedMaxDepth } =
+            expandAllTreePathData.viewer.editingContext.expandAllTreePath;
+          const newExpanded: string[] = [...expanded];
+
+          treeItemIdsToExpand?.forEach((itemToExpand) => {
+            if (!expanded.includes(itemToExpand)) {
+              newExpanded.push(itemToExpand);
+            }
+          });
+          setState((prevState) => ({
+            ...prevState,
+            expanded: newExpanded,
+            maxDepth: Math.max(expandedMaxDepth, maxDepth),
+          }));
+        }
+      }
+    }
+  }, [expandAllTreePathLoading, expandAllTreePathData]);
+
   const treeId: string = `modelBrowser://${leafType}?ownerKind=${encodeURIComponent(
     ownerKind
   )}&targetType=${encodeURIComponent(referenceKind)}&ownerId=${ownerId}&descriptionId=${encodeURIComponent(
     descriptionId
   )}&isContainment=${isContainment}`;
   const { tree } = useModelBrowserSubscription(editingContextId, treeId, state.expanded, state.maxDepth);
 
-  const onExpandedElementChange = (expanded: string[], maxDepth: number) => {
-    setState((prevState) => ({ ...prevState, expanded, maxDepth }));
+  const onExpand = (id: string, depth: number) => {
+    const { expanded, maxDepth } = state;
+
+    if (expanded.includes(id)) {
+      const newExpanded = [...expanded];
+      newExpanded.splice(newExpanded.indexOf(id), 1);
+
+      setState((prevState) => ({
+        ...prevState,
+        expanded: newExpanded,
+        maxDepth: Math.max(maxDepth, depth),
+      }));
+    } else {
+      setState((prevState) => ({ ...prevState, expanded: [...expanded, id], maxDepth: Math.max(maxDepth, depth) }));
+    }
   };
 
+  const onExpandAll = (treeItem: GQLTreeItem) => {
+    if (tree) {
+      const variables: GQLGetExpandAllTreePathVariables = {
+        editingContextId,
+        treeId: tree.id,
+        treeItemId: treeItem.id,
+      };
+      getExpandAllTreePath({ variables });
+    }
+  };
+
+  useEffect(() => {
+    if (tree?.id) {
+      const variables: GQLGetTreePathVariables = {
+        editingContextId,
+        treeId: tree.id,
+        selectionEntryIds: selectedTreeItemIds,
+      };
+      getTreePath({ variables });
+    }
+  }, [editingContextId, tree?.id, selectedTreeItemIds, getTreePath]);
+
+  useEffect(() => {
+    if (!treePathLoading) {
```

Comment:
We can remove at least two "if" from this function without a change in behavior, it would improve the readability

Reply:
Done